### ISSUE TITLE
StreamPipeReader NullReferenceException in some multithreading scenarios

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#101893]
### Description

NullReferenceException has been observed from StreamPipeReader. Likely caused by misuse (multiple concurrent calls to `ReadAsync`), however a more useful error should be thrown the NRE.

---

From investigation:

There is a nullref if the internal Stream.ReadAsync call returns and StreamPipeReader is adding a new BufferSegment while at the same time the reader is being completed and unsets BufferSegment state.

The specific lines are [BufferSegment.cs](https://github.com/dotnet/runtime/blob/c5c7f0d3d11cc82eddf1747fbdcaec9cb850c3aa/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs#L119) where segment.NextSegment can be null if [BufferSegment.cs](https://github.com/dotnet/runtime/blob/c5c7f0d3d11cc82eddf1747fbdcaec9cb850c3aa/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs#L69) sets _next to null right after the while check in SetNext.

Another bug in the same scenario is an ArgumentOutOfRange exception when incrementing the End value at [StreamPipeReader.cs](https://github.com/dotnet/runtime/blob/c5c7f0d3d11cc82eddf1747fbdcaec9cb850c3aa/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs#L275) which calls Slice on the AvailableMemory which is now empty from the Complete call.

### Reproduction Steps

Possibly caused by multiple concurrent ReadAsync calls.

### Expected behavior

More informative error. Or no error. Should look at what other pipe readers do in this situation.

### Actual behavior

NullReferenceException thrown.

### Regression?

No

### Configuration

.NET 8

**Comment by JamesNK** [comment]
cc @BrennanConroy @adityamandaleeka

**Comment by danmoseley** [comment]
This has already shipped at this point, so not urgent for 10. Moving to 11.

@BrennanConroy I guess this is something we should fix at some point though?

**Comment by BrennanConroy** [comment]
Right, this is about giving nicer errors and better detection for buggy app code.

**Comment by eerhardt** [comment]
I assume this can be moved out of .NET 11. Thoughts @BrennanConroy @JamesNK?
